### PR TITLE
feature(hardware): add mechanism that raises exception if a firmware update is required.

### PR DESCRIPTION
**Pull request by vegano1**
# Overview

If the OT3 requires an update to one of the firmware submodules (head, gantry, pipettes, etc) then we dont want the user to be able to use the robot until the updates are all complete. We want this to insure uniformity between releases so we know what a submodule should be running.

This will add an internal flag and decorator mechanism to be used for functions that send can messages, while we only want firmware updates and deviceInfoRequests/Response messages to go out. This decorator currently always defaults to False so nothing happens, but the next ticket will drive this based on whether a firmware update is required or not.

Closes [RCORE-475](https://opentrons.atlassian.net/browse/RCORE-475)
Merge after [RCORE-476](https://github.com/Opentrons/opentrons/pull/12035)

# Test Plan
- [x] Make sure the robot server service starts up on the OT3
- [x] Make sure this change does not affect functionality in any way, as it always defaults to false for now.

# Changelog

- Added requires_update decorator along with an internal flag that raises FirmwareUpdateRequired if the internal flag is True. NOTE: This is always false for now, will change once we add the logic in the next pr.
- Added requires_update decorator to functions in the OT3Controller class which sends can messages.
- Moved errors in opentrons.hardware_control.types into their own file opentrons.hardware_control.errors

# Review requests

- Any reason why we would not want to raise an exception here?

# Risk assessment

Low, OT3 only, and functionality is disabled by default

[RCORE-475]: https://opentrons.atlassian.net/browse/RCORE-475?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RCORE-476]: https://opentrons.atlassian.net/browse/RCORE-476?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: hardware/opentrons_hardware/firmware_update/errors.py [hunk 1]
```diff
@@ -34,3 +34,16 @@ class BootloaderNotReady(FirmwareUpdateException):
     """Bootloader is not ready."""
 
     pass
+
+
+class FirmwareUpdateRequired(FirmwareUpdateException):
```
Comment (sfoster1) [inline hardware/opentrons_hardware/firmware_update/errors.py]: I think this wants to live in `hardware_control/types.py` (or split that file into `hardware_control/types.py` and `hardware_control/errors.py`). This exception will have to be handled by layers above the `opentrons` module, and we don't want them to have to import this module to get at the exception class.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
The decorator looks good. I think we should move where the exception is defined, and we should drop in a test to make sure the decorator raises an exception properly.

**Review by sfoster1 — APPROVED** [review]
Looks great to me!

**Review by fsinapi — APPROVED** [review]
Nice!

**Pull request merged**